server/engine/dryrun.go

```diff
@@ -0,0 +1,204 @@
+package engine
+
+import (
+	"context"
+	"server/config"
+	"server/model"
+	"server/persistence"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/google/uuid"
+	"github.com/microsoft/commercial-marketplace-offer-deploy/pkg/api"
+	"github.com/microsoft/commercial-marketplace-offer-deploy/pkg/events"
+	"github.com/microsoft/commercial-marketplace-offer-deploy/pkg/operation"
+	"github.com/microsoft/commercial-marketplace-offer-deploy/sdk"
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+var (
+	dryRunInstance     *dryRunController
+	dryRunInstanceOnce sync.Once
+)
+
+type ErrorHandler func(err error)
+type NameOrKeyCreate func() string
+
+type dryRunController struct {
+	// the MODM deployment id
+	deploymentId   int
+	db             *gorm.DB
+	done           chan struct{}
+	clientEndpoint string
+	location       string
+	resourceGroup  string
+	subscription   string
+	apiKey         string
+	hookName       string
+	deploymentName string
+
+	// this is the url that will be called by MODM. It maps to /eventhook route for handler/eventhook
+	eventHookCallbackUrl string
+	HandleError          ErrorHandler
+}
+
+func (d *dryRunController) Execute(ctx context.Context) {
+	time.Sleep(10 * time.Second)
+
+	go func() {
+		step, err := d.getStep()
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		client, err := sdk.NewClient(d.clientEndpoint, cred, nil)
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		deploymentName := d.deploymentName
+		request := api.CreateDeployment{
+			Name:           &deploymentName,
+			Template:       step.Template,
+			Location:       &d.location,
+			ResourceGroup:  &d.resourceGroup,
+			SubscriptionID: &d.subscription,
+		}
+
+		dep, err := client.Create(ctx, request)
+		if err != nil {
+			d.HandleError(err)
+		}
+		d.deploymentId = int(*dep.ID)
+
+		createEventRequest := api.CreateEventHookRequest{
+			APIKey:   &d.apiKey,
+			Callback: &d.eventHookCallbackUrl,
+			Name:     &d.hookName,
+		}
+
+		_, err = client.CreateEventHook(ctx, createEventRequest)
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		res, err := client.DryRun(ctx, d.deploymentId, step.Parameters)
+		if err != nil {
+			d.HandleError(err)
+		}
+
+		d.create(res)
+	}()
+
+	<-d.done
+}
+
+func DryRunControllerInstance() *dryRunController {
+	dryRunInstanceOnce.Do(func() {
+		dryRunInstance = &dryRunController{
+			db:                   persistence.NewPersistentDB(config.GetEnvironment().DB_PATH).Instance,
```

Comment:
I think we could probably just add this to our existing DB, but probably not necessary to change now for this PR.

Reply:
To be done after merge